### PR TITLE
RavenDB-13902: linux vfork implementation, and initial RavenProcess c…

### PR DESCRIPTION
**Pull request by aviviadi**
…lass impl

### CODE REVIEW

**Comment by aviviadi** [comment]
This is draft PR (RavenServer.cs going to have different impl)
The PAL code is for review though

File: src/Raven.Pal/src/posix/linuxonly.c [hunk 1]
```diff
@@ -108,5 +108,14 @@ rvn_test_storage_durability(
     return rc;
 }
 
+PRIVATE int32_t
+rvn_pipe_for_child(int descriptors[])
```
Comment (ayende) [inline src/Raven.Pal/src/posix/linuxonly.c]:
```suggestion
rvn_pipe_for_child(int descriptors[2])
```

Or `int * descriptors` ?

File: src/Raven.Pal/src/posix/maconly.c [hunk 2]
```diff
@@ -68,4 +69,19 @@ rvn_test_storage_durability(
     return SUCCESS; /* windows and mac are always true */
 }
 
+PRIVATE int32_t
+rvn_pipe_for_child(int descriptors[])
```
Comment (ayende) [inline src/Raven.Pal/src/posix/maconly.c]:
```suggestion
rvn_pipe_for_child(int descriptors[2])
```

File: src/Raven.Pal/src/posix/spawn.c [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+parse_cmd_line(char* line)
+{
+    if (line == NULL || *line == '\0')
+        return NULL;
+
+    char** argv = malloc(strlen(line)*2);
+    char * pch;
+    int argc = 1;    
+    
+    pch = strtok (line," \t\n");
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: This mutates the `line`, is that expected?

File: src/Raven.Pal/src/posix/spawn.c [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+#define MSEC_TO_SEC(x) (x/1000)
+#define MSEC_TO_NANOSEC(x) ((x % 1000) * 1000000)
+
+PRIVATE char** 
+parse_cmd_line(char* line)
+{
+    if (line == NULL || *line == '\0')
+        return NULL;
+
+    char** argv = malloc(strlen(line)*2);
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: We don't actually need this size, do we? Let's default to `malloc(sizeof(char**) * 8)` and `realloc` as needed?

File: src/Raven.Pal/src/posix/spawn.c (hunk 3 above)
Comment (ayende) [reply src/Raven.Pal/src/posix/spawn.c]: Can we use `strtok_r`, instead?

File: src/Raven.Pal/src/posix/spawn.c [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        else
+        {
+            return -1;
+        }
+    }
+    return rv;
+}
+
+PRIVATE void 
+CloseIfOpen(int fd)
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: naming convention

File: src/Raven.Pal/src/posix/spawn.c [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+{
+    if (fd >= 0)    
+        close(fd);    
+}
+
+PRIVATE ssize_t WriteSize(int fd, const void* buffer, size_t count)
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: naming convention

File: src/Raven.Pal/src/posix/spawn.c [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+{
+    ssize_t rv = 0;
+    while (count > 0)
+    {
+        ssize_t result = write(fd, buffer, count);
+        while ( result  < 0 && errno == EINTR)
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: can we make this one liner, like in the pipe open?

File: src/Raven.Pal/src/posix/spawn.c [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        else
+        {
+            return -1;
+        }
+    }
+    return rv;
+}
+
+PRIVATE void 
+ExitChild(int pipeToParent, int error)
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: naming convention

File: src/Raven.Pal/src/posix/spawn.c [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+EXPORT int32_t
+rvn_spawn_process(const char* filename, char* cmdline, void** pid, void** standard_in, void** standard_out, int32_t* detailed_error_code) {    
+    int stdinFds[2] = {-1, -1}, stdoutFds[2] = {-1, -1}, waitForChildToExecPipe[2] = {-1, -1};
+    pid_t processId = -1;
+    int thread_cancel_state;
+    sigset_t signal_set;
+    sigset_t old_signal_set;
+
+    char **argv = parse_cmd_line(cmdline);
+    argv[0] = (char *)filename;
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: I don't like this, better to pass `file` to `parse_cmd_line` and have it set it there.

File: src/Raven.Pal/src/posix/spawn.c [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                void (*oldhandler)(int) = (sa_old.sa_flags & SA_SIGINFO) ? (void (*)(int))sa_old.sa_sigaction : sa_old.sa_handler;
+                if (oldhandler != SIG_IGN && oldhandler != SIG_DFL)
+                    sigaction(sig, &sa_default, NULL);
+            }
+        }
+        pthread_sigmask(SIG_SETMASK, &old_signal_set, &junk_signal_set); 
+        
+        /* redirect both stdout and stderr to the same stream, and redirect stdin: */
+        result = dup2(stdoutFds[1], STDOUT_FILENO);
+        while ( result < 0 && errno == EINTR)
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: can use one line here?

File: src/Raven.Pal/src/posix/spawn.c [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+
+    if (rc != SUCCESS)
+    {
+        CloseIfOpen(stdinFds[1]);
+        CloseIfOpen(stdoutFds[0]);
+        
+        if (processId > 0)
+        {
+            int status;
+            waitpid(processId, &status, 0);
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c:235]: can this cause us to hang?

File: src/Raven.Pal/src/posix/spawn.c [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+    sigemptyset(&child_mask);
+    sigaddset(&child_mask, SIGCHLD);
+    time_t timeout_sec = MSEC_TO_SEC(timeout_ms);
+    
+    int status;
+
+    time_t start = time(NULL);
+
+    struct timespec houndrad_ms;
+    houndrad_ms.tv_sec = 0;
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: why is this zero here?

File: src/Raven.Pal/src/posix/spawn.c [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+    houndrad_ms.tv_nsec = MSEC_TO_NANOSEC(timeout_ms);
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: If the value is 1000 ms, then this would be zero too, no?

File: src/Raven.Pal/src/posix/spawn.c [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+    sigset_t child_mask;
+    sigemptyset(&child_mask);
+    sigaddset(&child_mask, SIGCHLD);
+    time_t timeout_sec = MSEC_TO_SEC(timeout_ms);
+    
+    int status;
+
+    time_t start = time(NULL);
+
+    struct timespec houndrad_ms;
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: what is the meaning of the variable name?

File: src/Raven.Pal/src/posix/spawn.c [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+    houndrad_ms.tv_sec = 0;
+    houndrad_ms.tv_nsec = MSEC_TO_NANOSEC(timeout_ms);
+
+    struct timespec remaining;
+    /*busy wait for proc to end, can't use 'sigtimedwait' on OS X*/
```
Comment (ayende) [inline src/Raven.Pal/src/posix/spawn.c]: Why are we setting up signals above?

File: src/Raven.Pal/src/win/spawn.c [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+#include "status_codes.h"
+#include <stdio.h>
+
+
+
+EXPORT int32_t
+rvn_kill_process(void* pid, int32_t* detailed_error_code) {
+    if (!TerminateProcess(pid, 9)) {
+        *detailed_error_code = GetLastError();
+        return FAIL_KILL;
```
Comment (ayende) [inline src/Raven.Pal/src/win/spawn.c]: leaking the handle here, no?

File: src/Raven.Pal/src/win/spawn.c [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+    DWORD rc = WaitForSingleObject(pid, timeout_ms);
+    
+    if (rc == WAIT_TIMEOUT)
+    {
+        *detailed_error_code = GetLastError();
+        return FAIL_TIMEOUT;
+    }
+
+    if (!GetExitCodeProcess(pid, exit_code)) {
+        *detailed_error_code = GetLastError();
```
Comment (ayende) [inline src/Raven.Pal/src/win/spawn.c:33]: leaking the handle here, no?

File: src/Raven.Pal/src/win/spawn.c [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+success:
+    if (stdin_read != INVALID_HANDLE_VALUE)
+        CloseHandle(stdin_read);
+    if (stdout_write != INVALID_HANDLE_VALUE)
+        CloseHandle(stdout_write);
+
+    if(line != NULL)
+        free(line);
+    if (pi.hThread != INVALID_HANDLE_VALUE)
```
Comment (ayende) [inline src/Raven.Pal/src/win/spawn.c:143]: If we had an error before `CreateProcess`, then `pi` is zeroed, and `hThread` would be zero, so we'll try to close handle 0?
Let's set it explicitly above, before any failure can happen

**Review by ayende — CHANGES_REQUESTED** [review]
Didn't look at the C# code

File: src/Raven.Pal/src/posix/spawn.c (hunk 3 above)
Comment (aviviadi) [reply src/Raven.Pal/src/posix/spawn.c]: strtok_r is not thread safe. If someone passes the same string twice (tests?) then we should avoid the _r func.

Comment (ayende) [reply src/Raven.Pal/src/posix/spawn.c]: That is the other way around, actually

**Comment by aviviadi** [comment]
comments fixed in https://github.com/ravendb/ravendb/pull/9723

**Pull request closed without merging**